### ISSUE TITLE
How to retrieve target architecture using defines?

### ISSUE THREAD

**Issue by EliteMasterEric** [openfl/lime#1486]
The following XML tags function as expected:

```
	<echo value="Platform architecture: 32-bit" if="HXCPP_M32" />
	<echo value="Platform architecture: 64-bit" if="HXCPP_M64" />
```

However, there is an issue with this; if the user has omitted these compile flags and simply uses the default `lime build linux` command (which builds for their current architecture), the result is that **neither of these flags will be true**.

Is there a way to print when the user is, say, building for x64, then the user has not explicitly set the `HXCPP_M64` flag?

**Comment by player-03** [comment]
Ok, give my pull request a try, and give [include.hxp](https://lime.software/docs/project-files/hxp-format/) a shot. You're interested in the [`architectures` variable](https://github.com/haxelime/lime/blob/develop/src/lime/tools/HXProject.hx#L27), which will contain all the default architectures for this platform.

You can also check the [`defines` variable](https://github.com/haxelime/hxp/blob/master/src/hxp/Script.hx#L15) for `HXCPP_M32` or `HXCPP_M64` in case the user set those by hand.

**Comment by MasterEric** [comment]
@player-03 I didn't test your PR since it's only for haxelibs, but when I built a `project.hxp` and ran it I got this output:

```
E:\Program Files\HaxeToolkits\424\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:53: characters 12-29 : Warning : Std.is is deprecated. Use Std.isOfType instead.
E:\Program Files\HaxeToolkits\424\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:58: characters 12-28 : Warning : Std.is is deprecated. Use Std.isOfType instead.
E:\Program Files\HaxeToolkits\424\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:78: characters 12-32 : Warning : Std.is is deprecated. Use Std.isOfType instead.
E:\Program Files\HaxeToolkits\424\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:94: characters 12-29 : Warning : Std.is is deprecated. Use Std.isOfType instead.
E:\Program Files\HaxeToolkits\424\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:110: characters 12-27 : Warning : Std.is is deprecated. Use Std.isOfType instead.
Called from lime/_internal/backend/native/NativeCFFI.hx line 609
Called from lime/system/CFFI.hx line 172
Called from lime/system/CFFI.hx line 306
Uncaught exception - Could not find NekoAPI interface.
```

**Comment by player-03** [comment]
The deprecation warnings are annoying but not the problem. They'll be fixed whenever the next release happens to be, but if they're annoying in the meantime, the error message is right about how to fix them. (I think you'll then need to run `lime rebuild tools`.)

As for "Could not find NekoAPI interface," that means it couldn't find `lime.ndll`. Which is strange, because if you've previously compiled using project.xml, then the file had to exist. Well ok, it's simple enough to check.

1. Run `haxelib path lime` to make sure it prints the correct folder.
2. Search Lime's ndll directory (specifically your system's folder, probably ndll/Windows?) for `lime.ndll`.

**Comment by MasterEric** [comment]
> The deprecation warnings are annoying but not the problem.

I thought as much, just wanted to paste the whole log for the sake of completeness.

> Search Lime's ndll directory (specifically your system's folder, probably ndll/Windows?) for lime.ndll.

`lime.ndll` is present in both `C:\HaxeToolkit\haxe\lib\lime\7,9,0\ndll\Windows64` and `C:\HaxeToolkit\haxe\lib\lime\7,9,0\ndll\Windows`.

**Comment by player-03** [comment]
> `lime.ndll` is present in both `C:\HaxeToolkit\haxe\lib\lime\7,9,0\ndll\Windows64` and `C:\HaxeToolkit\haxe\lib\lime\7,9,0\ndll\Windows`.

Figures. Well, in that case I'm not sure what's going on, so you're going to have to do a little more digging.

I'd say the next step is to comment out [this `try-catch` statement](https://github.com/haxelime/lime/blob/develop/src/lime/system/CFFI.hx#L283). The true error must originate [somewhere in `load()`](https://github.com/haxelime/lime/blob/develop/src/lime/system/CFFI.hx#L96-L164), but since it gets caught, we don't know exactly where.

**Comment by MasterEric** [comment]
```
Called from lime/_internal/backend/native/NativeCFFI.hx line 609
Called from lime/system/CFFI.hx line 172
Called from lime/system/CFFI.hx line 286
Called from lime/system/CFFI.hx line 111
Called from C:\HaxeToolkit\haxe\std/neko/Lib.hx line 35
Uncaught exception - load.c(237) : Failed to load library : C:\HaxeToolkit\haxe\lib\lime/7,9,0/ndll/Windows/lime.ndll
```

Not a helpful message, to be sure.

**Comment by player-03** [comment]
I mean, even though it doesn't tell us how to solve the problem, it's still worth eliminating possibilities.

(It does point us to a [much more specific line](https://github.com/HaxeFoundation/neko/blob/v2-3-0/vm/load.c#L237). Looking at that function, we can see that `dlopen()` failed on line 226, but thanks to line 231 it doesn't print the error message. Sadly, it's infeasible to modify `load.c`, so we can't directly fix this.)

Maybe the solution is as simple as reinstalling Neko, but I doubt that too. Neko normally works, right? You can target it?

```haxe
openfl create DisplayingABitmap
cd DisplayingABitmap
lime test neko
```

Let's look at this from another angle. Lime always uses Neko to build projects. You can reliably build using project.xml, but not project.hxp. The main difference with project.hxp is, it recursively calls both [Haxe](https://github.com/haxelime/lime/blob/7.9.0/src/lime/tools/HXProject.hx#L578) and [Neko](https://github.com/haxelime/lime/blob/7.9.0/src/lime/tools/HXProject.hx#L605). Which ought to be fine, but apparently isn't.

Try adding `trace(args)` on line 576, `trace(inputData)` on line 601, and `trace("recursion done")` on line 615. The first will show up, the second might, and the third won't. (Remember `lime rebuild tools`, but I assume you know that.)

Depending on your results, I'll have some other tests to suggest.

(If this is getting to be too much work, you could try a workaround. But be aware it wouldn't be future-proof.)

**Comment by MasterEric** [comment]
* I'm not sure what the Neko platform is as I've never used it for building before. Can you give a brief rundown of what it is and why it's used?
* Running `lime test neko` on the DisplayingABitmap project works (and builds really quickly). Running it on my XML project returns an error due to its use of the cpp package.
* Adding the trace calls you suggested provides a very odd result; NONE of them appear when using the HXP project, using the `windows` or `neko` platform, even when doing a clean build.

> If this is getting to be too much work, you could try a workaround

Not a problem for me at all. I have my project setup so that I can continue working on it by renaming the project to `project.hxpold` which will revert back to the XML.

**Comment by player-03** [comment]
> * I'm not sure what the Neko platform is as I've never used it for building before. Can you give a brief rundown of what it is and why it's used?

Neko is a programming language/virtual machine. I think it was originally designed as its own standalone language (and [its website](https://nekovm.org/) pitches it as such), but it quickly turned into nothing more than "Haxe's rapid testing target."

Pros: as you noticed, it builds really quickly. Cons: it isn't super fast, and no one installs the VM unless they're a Haxe developer. You'd never distribute Neko bytecode to anyone besides a Haxe developer.

But you certainly can distribute Neko bytecode to Haxe developers. In fact, that's a big part of Haxelib. When you call `haxelib run lime` (or any other library), it runs that library's "run.n" file - a Neko file. `lime rebuild tools` rebuilds run.n (plus other things sometimes, but mostly run.n). Lime's run.n includes all the code to parse both project.xml and project.hxp.

Interestingly, Haxelib never checks for run.hl, even though [HashLink](https://hashlink.haxe.org/) is supposed to replace Neko.

> * Running `lime test neko` on the DisplayingABitmap project works (and builds really quickly). Running it on my XML project returns an error due to its use of the cpp package.

To be expected, but still nice to know.

> * Adding the trace calls you suggested provides a very odd result; NONE of them appear when using the HXP project, using the `windows` or `neko` platform, even when doing a clean build.

And you made sure to run `lime rebuild tools`? Because without that, your changes won't make it into run.n.

**Comment by MasterEric** [comment]
> Interestingly, Haxelib never checks for run.hl, even though HashLink is supposed to replace Neko.

Thanks for the explainer on this. I don't plan on using Neko in the future since it doesn't appear to have a debugger in VSCode. I tried HashLink in the past but when I run `lime rebuild` for one of my libraries it creates an NDLL instead of an HDLL (but that's its own issue).

> And you made sure to run `lime rebuild tools`?

I did not, nice catch. Here's the new output:

```
../src/lime/system/CFFI.hx:285: Reached problem method (hi player-03)
../src/lime/tools/HXProject.hx:576: ARGS: [Project,-main,lime.tools.HXProject,-cp,C:\Users\Eric\AppData\Local\Temp/temp_16432019,-neko,C:\Users\Eric\AppData\Local\Temp/temp_16432019/Project.n,-cp,C:/HaxeToolkit/haxe/lib/hxp/1,2,1/src,-lib,lime,-lib,hxp]
C:\HaxeToolkit\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:53: characters 12-29 : Warning : Std.is is deprecated. Use Std.isOfType instead.
C:\HaxeToolkit\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:58: characters 12-28 : Warning : Std.is is deprecated. Use Std.isOfType instead. 
C:\HaxeToolkit\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:78: characters 12-32 : Warning : Std.is is deprecated. Use Std.isOfType instead. 
C:\HaxeToolkit\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:94: characters 12-29 : Warning : Std.is is deprecated. Use Std.isOfType instead. 
C:\HaxeToolkit\haxe\lib\hxp/1,2,1/src/hxp/ObjectTools.hx:110: characters 12-27 : Warning : Std.is is deprecated. Use Std.isOfType instead.
../src/lime/tools/HXProject.hx:601: INPUTDATA: oy13:processDryRunfy4:namey7:Projecty14:logEnableColorty5:debugfy10:logVerbosefy11:userDefinesbhy7:commandy5:buildy6:targety7:windowsy11:targetFlagsbhy11:environmentby13:NEKO_INSTPATHy25:C%3A%5CHaxeToolkit%5Cnekoy5:WXWINy22:E%3A%5CwxWidgets-3.1.5y12:TERM_PROGRAMy6:vscodey9:COLORTERMy9:truecolory8:haxe_very5:4.2.4y10:USERDOMAINy15:DESKTOP-1GMV4AIy13:ANDROID_SETUPy4:truey12:LOCALAPPDATAy37:C%3A%5CUsers%5CEric%5CAppData%5CLocaly22:PROCESSOR_ARCHITECTUREy5:AMD64y21:ProgramFiles%28x86%29y34:C%3A%5CProgram%20Files%20%28x86%29y16:ANDROID_NDK_ROOTy17:E%3A%5CAndroidNDKy31:FPS_BROWSER_USER_PROFILE_STRINGy7:Defaulty11:USERPROFILEy19:C%3A%5CUsers%5CEricy20:NUMBER_OF_PROCESSORSy1:8y7:HL_PATHy29:C%3A%5CHaxeToolkit%5CHashLinky4:TEMPy44:C%3A%5CUsers%5CEric%5CAppData%5CLocal%5CTempy18:CommonProgramFilesy39:C%3A%5CProgram%20Files%5CCommon%20Filesy12:COMPUTERNAMER24y13:HAXE_STD_PATHy34:C%3A%5CHaxeToolkit%5Chaxe%5Cstd%5Cy12:ProgramW6432y22:C%3A%5CProgram%20Filesy9:HOMEDRIVEy4:C%3Ay10:SystemRooty14:C%3A%5CWindowsy25:CHROME_CRASHPAD_PIPE_NAMEy47:%5C%5C.%5Cpipe%5Ccrashpad_7048_SZJGDIWUKCEBGPRMy20:TERM_PROGRAM_VERSIONy6:1.62.0y16:HAXELIB_RUN_NAMEy4:limey11:HAXELIB_RUNy1:1y6:PUBLICy21:C%3A%5CUsers%5CPublicy4:PATHy1014:E%3A%5CProgram%20Files%5CJava%5Cjdk1.8.0_301%2Fbin%3BC%3A%5CProgram%20Files%20%28x86%29%5CCommon%20Files%5COracle%5CJava%5Cjavapath%3BC%3A%5CWindows%5Csystem32%3BC%3A%5CWindows%3BC%3A%5CWindows%5CSystem32%5CWbem%3BC%3A%5CWindows%5CSystem32%5CWindowsPowerShell%5Cv1.0%5C%3BC%3A%5CWindows%5CSystem32%5COpenSSH%5C%3BC%3A%5CProgram%20Files%5CVS%20Code%5Cbin%3BC%3A%5CProgram%20Files%5CGit%5Ccmd%3BE%3A%5CProgram%20Files%5Cnodejs%5C%3BE%3A%5CProgram%20Files%5Cffmpeg%5Cbin%3BC%3A%5CProgram%20Files%5Cdotnet%5C%3BC%3A%5CHaxeToolkit%5Cneko%3BC%3A%5CHaxeToolkit%5Chaxe%3BC%3A%5CHaxeToolkit%5Cneko%3BC%3A%5CUsers%5CEric%5CAppData%5CLocal%5CPrograms%5CPython%5CPython39%5CScripts%5C%3BC%3A%5CUsers%5CEric%5CAppData%5CLocal%5CPrograms%5CPython%5CPython39%5C%3BC%3A%5CUsers%5CEric%5CAppData%5CLocal%5CMicrosoft%5CWindowsApps%3BC%3A%5CUsers%5CEric%5CAppData%5CLocal%5Cgitkraken%5Cbin%3BC%3A%5CUsers%5CEric%5CAppData%5CRoaming%5Cnpm%3BC%3A%5CHaxeToolkit%5Chaxe%3BC%3A%5CHaxeToolkit%5Cneko%3BC%3A%5CUsers%5CEric%5C.dotnet%5Ctoolsy11:GIT_ASKPASSy91:c%3A%5CProgram%20Files%5CVS%20Code%5Cresources%5Capp%5Cextensions%5Cgit%5Cdist%5Caskpass.shy8:NEKOPATHR14y20:PROCESSOR_IDENTIFIERy67:Intel64%20Family%206%20Model%20158%20Stepping%209%2C%20GenuineIntely18:CommonProgramW6432R46y28:ORIGINAL_XDG_CURRENT_DESKTOPy9:undefinedy2:OSy10:Windows_NTy7:PATHEXTy80:.COM%3B.EXE%3B.BAT%3B.CMD%3B.VBS%3B.VBE%3B.JS%3B.JSE%3B.WSF%3B.WSH%3B.MSC%3B.CPLy21:VSCODE_GIT_IPC_HANDLEy43:%5C%5C.%5Cpipe%5Cvscode-git-63efe67b81-socky18:PROCESSOR_REVISIONy4:9e09y7:APPDATAy39:C%3A%5CUsers%5CEric%5CAppData%5CRoamingy4:LANGy11:en_US.UTF-8y12:PSModulePathy166:D%3A%5CDocuments%5CWindowsPowerShell%5CModules%3BC%3A%5CProgram%20Files%5CWindowsPowerShell%5CModules%3BC%3A%5CWindows%5Csystem32%5CWindowsPowerShell%5Cv1.0%5CModulesy23:VSCODE_GIT_ASKPASS_MAINy96:c%3A%5CProgram%20Files%5CVS%20Code%5Cresources%5Capp%5Cextensions%5Cgit%5Cdist%5Caskpass-main.jsy11:LOGONSERVERy21:%5C%5CDESKTOP-1GMV4AIy11:ProgramDatay18:C%3A%5CProgramDatay5:haxe4R63y11:SystemDriveR53y3:TMPR44y11:ANDROID_SDKy17:E%3A%5CAndroidSDKy11:SESSIONNAMEy7:Consoley23:VSCODE_GIT_ASKPASS_NODEy45:C%3A%5CProgram%20Files%5CVS%20Code%5CCode.exey4:Pathy961:C%3A%5CProgram%20Files%20%28x86%29%5CCommon%20Files%5COracle%5CJava%5Cjavapath%3BC%3A%5CWindows%5Csystem32%3BC%3A%5CWindows%3BC%3A%5CWindows%5CSystem32%5CWbem%3BC%3A%5CWindows%5CSystem32%5CWindowsPowerShell%5Cv1.0%5C%3BC%3A%5CWindows%5CSystem32%5COpenSSH%5C%3BC%3A%5CProgram%20Files%5CVS%20Code%5Cbin%3BC%3A%5CProgram%20Files%5CGit%5Ccmd%3BE%3A%5CProgram%20Files%5Cnodejs%5C%3BE%3A%5CProgram%20Files%5Cffmpeg%5Cbin%3BC%3A%5CProgram%20Files%5Cdotnet%5C%3BC%3A%5CHaxeToolkit%5Cneko%3BC%3A%5CHaxeToolkit%5Chaxe%3BC%3A%5CHaxeToolkit%5Cneko%3BC%3A%5CUsers%5CEric%5CAppData%5CLocal%5CPrograms%5CPython%5CPython39%5CScripts%5C%3BC%3A%5CUsers%5CEric%5CAppData%5CLocal%5CPrograms%5CPython%5CPython39%5C%3BC%3A%5CUsers%5CEric%5CAppData%5CLocal%5CMicrosoft%5CWindowsApps%3BC%3A%5CUsers%5CEric%5CAppData%5CLocal%5Cgitkraken%5Cbin%3BC%3A%5CUsers%5CEric%5CAppData%5CRoaming%5Cnpm%3BC%3A%5CHaxeToolkit%5Chaxe%3BC%3A%5CHaxeToolkit%5Cneko%3BC%3A%5CUsers%5CEric%5C.dotnet%5Ctoolsy15:ALLUSERSPROFILER95y10:DriverDatay48:C%3A%5CWindows%5CSystem32%5CDrivers%5CDriverDatay30:FPS_BROWSER_APP_PROFILE_STRINGy19:Internet%20Explorery8:USERNAMEy4:Ericy8:OneDrivey30:C%3A%5CUsers%5CEric%5COneDrivey27:CommonProgramFiles%28x86%29y51:C%3A%5CProgram%20Files%20%28x86%29%5CCommon%20Filesy8:HAXEPATHy25:C%3A%5CHaxeToolkit%5Chaxey7:ComSpecy35:C%3A%5CWindows%5Csystem32%5Ccmd.exey15:PROCESSOR_LEVELy1:6y6:windirR55y4:haxeR22y12:ProgramFilesR51y8:HOMEPATHy15:%5CUsers%5CEricy6:PROMPTy8:%24P%24Gy25:USERDOMAIN_ROAMINGPROFILER24y9:JAVA_HOMEy44:E%3A%5CProgram%20Files%5CJava%5Cjdk1.8.0_301hy12:haxelibDebugfy13:templatePathsahg
lime/system/CFFI.hx:285: Reached problem method (hi player-03)
Called from lime/_internal/backend/native/NativeCFFI.hx line 609
Called from lime/system/CFFI.hx line 172
Called from lime/system/CFFI.hx line 286
Called from lime/system/CFFI.hx line 111
Called from C:\HaxeToolkit\haxe\std/neko/Lib.hx line 35
Uncaught exception - load.c(237) : Failed to load library : C:\HaxeToolkit\haxe\lib\lime/7,9,0/ndll/Windows/lime.ndll
```

**Comment by player-03** [comment]
> when I run `lime rebuild` for one of my libraries it creates an NDLL instead of an HDLL

Yeah, Haxelib isn't the only piece of software that still prefers Neko...

> `../src/lime/system/CFFI.hx:285: Reached problem method (hi player-03)`
> `lime/system/CFFI.hx:285: Reached problem method (hi player-03)`

Interesting to note the slight difference. Not sure if it's relevant, but if we can't find anything else, it'll be worth checking up on.

Also it proves yet again that the method _can_ work. The first time, it's fine.

> `../src/lime/tools/HXProject.hx:601: INPUTDATA: //...`

Ok, I'm a bit surprised to see this. I'd suspected the Haxe phase, but ok.

To review what's happening:

1. It locates the project.hxp file and does some preprocessing. Really basic stuff, just looking to see if there are any extra source paths.
2. It makes a copy with the .hx extension.
3. It runs Haxe to compile the temporary .hx file, outputting a temporary Neko file. (I had you print the arguments passed to Haxe.)
4. It collects some information the project may need, and saves it in a temporary .dat file. (I had you print this too.)
5. It runs the Neko file, which does a bunch of stuff and outputs a .dat in the end.
6. It loads the output back into the original process, then cleans up.

The error is happening in step 5. Since it doesn't reach 6, the only thing to be deleted is input.dat. I had you print the data so you could recreate it, but you could also just comment line 612.

Either way, the next step is to try running that Neko file from the command line:

```text
$ cd C:/path/to/temp/project_123456
$ Neko Project.n input.dat output.dat
```

If all goes according to plan, you should get the exact same error. Once you do, you can try possible ways to fix it. Though I can only think of one right now.

Maybe your temp folder is on a different drive, or partition, or something. This can mess with file access sometimes. Try moving the temporary files to somewhere you know works, like the main project folder, and try running Neko there.

If you can't fix it, see if you can at least produce other errors. Run the Neko file with bad arguments, or no arguments.

**Comment by MasterEric** [comment]
I moved the temporary generated project file out, and I get the same error trying to run it locally like so:

```
PS D:\Programming\Game Modding\Friday Night Funkin\EnigmaEngine\tmp\temp_4296021> neko ./Project.n
Called from lime/_internal/backend/native/NativeCFFI.hx line 609
Called from lime/system/CFFI.hx line 172
Called from lime/system/CFFI.hx line 286
Called from lime/system/CFFI.hx line 111
Called from C:\HaxeToolkit\haxe\std/neko/Lib.hx line 35
Uncaught exception - load.c(237) : Failed to load library : C:\HaxeToolkit\haxe\lib\lime/7,9,0/ndll/Windows/lime.ndll

PS D:\Programming\Game Modding\Friday Night Funkin\EnigmaEngine\tmp\temp_4296021> neko ./Project.n blablabla
Called from lime/_internal/backend/native/NativeCFFI.hx line 609
Called from lime/system/CFFI.hx line 172
Called from lime/system/CFFI.hx line 286
Called from lime/system/CFFI.hx line 111
Called from C:\HaxeToolkit\haxe\std/neko/Lib.hx line 35
Uncaught exception - load.c(237) : Failed to load library : C:\HaxeToolkit\haxe\lib\lime/7,9,0/ndll/Windows/lime.ndll

PS D:\Programming\Game Modding\Friday Night Funkin\EnigmaEngine\tmp\temp_4296021> neko ./Project.n input.dat output.dat
Called from lime/_internal/backend/native/NativeCFFI.hx line 609
Called from lime/system/CFFI.hx line 172
Called from lime/system/CFFI.hx line 286
Called from lime/system/CFFI.hx line 111
Called from C:\HaxeToolkit\haxe\std/neko/Lib.hx line 35
Uncaught exception - load.c(237) : Failed to load library : C:\HaxeToolkit\haxe\lib\lime/7,9,0/ndll/Windows/lime.ndll
```

**Comment by player-03** [comment]
I'm running dangerously low on easy-to-test ideas...

- Did you ever try reinstalling Neko?
- Does this happen in a fresh project like DisplayingABitmap?
- What if you comment line 111 of CFFI.hx? (Do **_not_** run `lime rebuild tools` afterwards.)

**Comment by MasterEric** [comment]
Hey @player-03, sorry for the delay in getting back to you on this.

Here is a test repository I made for working on this issue: https://github.com/MasterEric/Neko-DisplayingABitmap

It contains the DisplayingABitmap project, a basic HXP file, and the temporary Neko file the project created. Sadly the issue occurs in this repo as well.

For your ideas:

* I deleted the existing Neko installation and replaced it with a fresh version from the website with the same result.
* See above, the issue occurs in that repo as well.
* Replacing `load` results in no noticeable change.

The thing I'm wondering about is whether some configuration file is wrong, and it's unable to locate the relevant binary. I uninstalled and reinstalled Haxe at one point, maybe a path option or environment variable is wrong?

**Comment by player-03** [comment]
No worries about the delay. Develop at your own pace!

Thanks for the test repo. While I can compile the project fine myself, I do get an error when running the Neko file. Interestingly, though, mine looks like this:

```text
Called from lime/_internal/backend/native/NativeCFFI.hx line 609
Called from lime/system/CFFI.hx line 172
Called from lime/system/CFFI.hx line 286
Called from lime/system/CFFI.hx line 111
Called from C:\HaxeToolkit\haxe\std/neko/Lib.hx line 35
Uncaught exception - load.c(237) : Failed to load library : lime.ndll (lime.ndll: cannot open shared object file: No such file or directory)
```

Couple notable things about that last line. I got an error description, but [that's just because I'm not on Windows](https://github.com/HaxeFoundation/neko/blob/v2-3-0/vm/load.c#L231-L235). More interestingly, it didn't spell out a path to `lime.ndll`, which I guessed meant it was looking in the current working directory.

This is very wrong, of course, but it implies a workaround - copy lime.ndll into the folder next to Project.n.

When I did, I got a new error: `Could not find asset path "Assets"`. This is exactly what it says on the tin. project.hxp says to check the Assets/ folder for assets, but this directory has no such folder. The solution is to run Project.n from the project root folder, like Lime does:

```bash
$ neko Temp/Project.n Temp/input.dat Temp/output.dat
```

Which brought back the first error, because I forgot to move lime.ndll from Temp/ to the project root. After moving it, Neko exited successfully, producing an output.dat file. I didn't care to run it through `Unserializer`, but the point is, it's progress.

**tl;dr:** Try copying lime.ndll into your project root, see if that helps.

**Comment by MasterEric** [comment]
@player-03 By copying `lime.ndll` from `C:/HaxeToolkit/haxe/lib/lime/7,9,0/ndll/Windows64/` into the same folder as the `project.hxp` file, builds will now successfully create a valid executable!

🎉 

My only side effect is that [this issue](https://github.com/haxelime/lime-vscode-extension/issues/88) still occurs (but that happens regardless if I use HXP or not), so I have to disable info logging in my build scripts for now, but otherwise your solution works, and can help point us to the proper solution to solve the problem (through a change to lime itself or otherwise).

HXP is definitely a lot better than XML when it works. I like being able to use actual logic rather than a chain of `if` and `unless` tags.

**Comment by player-03** [comment]
Great! It's still unclear why this happens (maybe a quirk in Haxe 4.1.5?), but it works and that's what matters.

It'd still be nice to find a solution that doesn't involve copying this 8 MB file into every new project you make. For instance, try adding `C:/HaxeToolkit/haxe/lib/lime/7,9,0/ndll/Windows64/` to PATH.

**Comment by player-03** [comment]
Updating PATH didn't work?

Honestly, I know very little about dll linking. This isn't the only case where it's caused problems, and I just don't know how to debug it. Sorry.

We could also try the approach from #1615 - instead of compiling a Neko binary for the project, just use `--interp`. That neatly avoids all dll-related errors.

**Comment by player-03** [comment]
Ok, I updated that pull request. On that branch, `lime test <target> -eval` should build without needing lime.ndll.

**Comment by player-03** [comment]
Unfortunately, I think `-eval` will have to be the answer whenever lime.ndll fails. It's come up multiple times, and everyone was very helpful in testing, but we never found a better option. I'm guessing the bug lies somewhere in Neko's dll-loading code, which is no longer maintained.

Closing this because both solutions (one for project.xml, one for project.hxp) were merged.

**Comment by EliteMasterEric** [comment]
Valid. If I end up trying to use HXP again in the future and run into any issues, I will raise a new issue appropriately.

**Comment by tobil4sk** [comment]
I tried this out, and I was able to reproduce the issue with Neko 2.3.0. However, using the latest git version of Neko seems to resolve this issue, so if you end up using hxp again it might be worth trying if using an updated version of Neko fixes it.

There should be a 2.4.0 release of Neko soon, because there was a release needed for to packaging reasons. So hopefully it should be resolved in this release.
